#2604 Platform: Hosting en domeinnaam verhuizen

tfrijsewijk opened the pull request.

iterox commented on scripts/update-azure-blob-storage/main.ts:
```diff
@@ -0,0 +1,128 @@
+/**
+ * This file:
+ * - Updates dso-toolkit.nl/versions.json with all the deployed releases (tags and branches) in the Azure Blob Storage container.
```
Hoe werd dit voorheen geregeld? En waarom is dit nu nodig?

tfrijsewijk replied:
Vroeger werd dit in https://github.com/dso-toolkit/dso-toolkit/pull/2683/files#diff-30883d1df2fc62f59d66255a70dc05f60d3d4ef4d3957e6e3e38e62f6e471bd0L38 geregeld:

Er werd een SSH verbinding gestart waarna er een NPM script op de host werd gedraaid. Dit script voerde uit wat er nu in `main.ts` staat.

Ik zal het oude script in een comment van dit PR plakken :+1:

tfrijsewijk replied:
Eric zegt ✅

iterox reviewed (commented):
Ik heb een paar vragen gesteld.

tfrijsewijk commented on .github/workflows/ci.yml:
```diff
@@ -28,8 +28,8 @@ jobs:
               }
-              else if (type === 'heads' && ref === 'master') {
-                core.exportVariable('DT_REF', 'master');
+              else if (type === 'heads' && (ref === 'master' || ref === 'test')) {
```
Nee, ik heb dit ingebouwd zodat we kunnen testen tegen de test omgeving.

Na het mergen van deze branch (`test`) verwijder ik wel de branch. En als we dan aan het platform gaan werken doen we dat vanaf de branch `test`.

Al denkende, misschien kunnen we wel beter de conventie maken dat branches _beginnen_ met `test-` zodat we weer 'n branch per issue hebben.

tfrijsewijk replied:
Eric zegt ✅

tfrijsewijk commented on .github/workflows/ci.yml:
```diff
@@ -80,20 +80,18 @@ jobs:
-          DT_DEPLOY_PORT: ${{ secrets.DT_DEPLOY_PORT }}
-          DT_DEPLOY_USER: ${{ secrets.DT_DEPLOY_USER }}
-          DT_DEPLOY_ROOT: ${{ secrets.DT_DEPLOY_ROOT }}
+          DT_AZURE_STORAGE_ACCOUNT_NAME: ${{ env.DT_REF == 'test' && secrets.DT_AZURE_STORAGE_ACCOUNT_NAME_ACC || secrets.DT_AZURE_STORAGE_ACCOUNT_NAME }}
```
Nee, de secrets die naar `_ACC` verwijzen blijven en worden gebruikt zodra `DT_REF == "test"`.

We hebben een extra omgeving maar die heeft 'n beetje 'n identiteitscrisis: De ene keer wordt ie 'acc' en de andere keer 'test' genoemd.

Ik ga hier nog even met @Dutchy- over bomen...

Dutchy- replied:
De korte samenvatting is, jullie hebben een acceptatie-omgeving gekregen op ons 'ota' cluster. Dat cluster heet op bepaalde plekken 'test' en daarom zie je dat af en toe terug, onder andere in de url.

Als het echt moet kan ik dingen hernoemen van acc naar test. Andersom, van test naar acc, is niet mogelijk. Maar ik zou gewoon accepteren dat de omgeving een beetje een identiteitscrisis heeft en het daar bij laten.

tfrijsewijk replied:
Eric zegt ✅

tfrijsewijk commented:
De legacy `prune-topic-branches.js`:

```js
const axios = require("axios");
const path = require("path");
const fs = require("fs");

// Start configuration
const repository = "dso-toolkit/dso-toolkit";
const sites = [
  "angular.dso-toolkit.nl/www",
  "cdn.dso-toolkit.nl/www",
  "cdn.dso-toolkit.nl/www/dso-toolkit",
  "cdn.dso-toolkit.nl/www/@dso-toolkit/core",
  "core.dso-toolkit.nl/www",
  "css.dso-toolkit.nl/www",
  "dso-toolkit.nl/www",
  "react.dso-toolkit.nl/www",
  "storybook.dso-toolkit.nl/www",
];
const wwwRoot = path.resolve("..", "dso-toolkit.nl", "www");
// End configuration

axios
  .get(`https://api.github.com/repos/${repository}/branches`, {
    params: {
      per_page: 100,
    },
  })
  .then(({ data }) => {
    const branches = data.map((b) => ({
      ...b,
      directory: formatBranchToDirectory(b.name),
    }));

    pruneDirectories(branches);
    updateVersions();
  });

function pruneDirectories(branches) {
  for (const site of sites.map((s) => path.resolve("..", s))) {
    fs.readdirSync(site)
      .filter(
        (file) =>
          file[0] === "_" &&
          file[1] !== "_" &&
          fs.lstatSync(path.resolve(site, file)).isDirectory()
      )
      .filter((d) => !branches.some((b) => b.directory === d))
      .forEach((d) => {
        const from = path.resolve(site, d);

        console.log(`PRUNING ${from}`);

        if (!fs.existsSync(from)) {
          console.log(`- does not exist, skipping`);
        } else {
          for (let i = 0; true; i++) {
            const to = path.resolve(
              from,
              "../__bak/",
              d + (i > 0 ? `_${i}` : "")
            );

            if (!fs.existsSync(to)) {
              console.log(`- Moving ${d} from ${from} to ${to}`);
              fs.renameSync(from, to);

              break;
            }
          }
        }
      });
  }
}

function updateVersions() {
  const refs = fs
    .readdirSync(wwwRoot)
    .filter((file) => fs.lstatSync(path.resolve(wwwRoot, file)).isDirectory())
    .filter((directory) => isSelectableBranch(directory))
    .reduce(
      (list, directory) => {
        if (directory[0] === "_") {
          list.branches.push(directory);
        } else if (isSemVer(directory, true)) {
          list.versions.push(directory);
        }

        return list;
      },
      { versions: [], branches: [] }
    );

  const versions = [
    ...refs.versions.sort(semverCompare),
    "master",
    "latest",
    "next",
    ...refs.branches,
  ];

  const latest = refs.versions.filter((version) => isSemVer(version)).pop();

  console.log(`Latest version: ${latest}`);

  const versionsJson = JSON.stringify(
    { versions: versions.map(getVersion) },
    null,
    2
  );

  fs.writeFileSync(path.resolve(wwwRoot, "versions.json"), versionsJson);

  const versionMatch = '<meta http-equiv="refresh" content="0; url=/(.+)" />';

  const indexHtml = fs
    .readFileSync(path.resolve(wwwRoot, "index.html"))
    .toString();
  const regex = new RegExp(versionMatch);

  fs.writeFileSync(
    path.resolve(wwwRoot, "index.html"),
    indexHtml.replace(regex, versionMatch.replace("(.+)", latest))
  );
}

function formatBranchToDirectory(branch) {
  if (branch[0] === "#") {
    return `_${branch.substr(1)}`;
  }

  return branch;
}

function isSelectableBranch(directory) {
  return (
    directory === "master" ||
    directory === "next" ||
    directory === "latest" ||
    (directory[0] === "_" && directory[1] !== "_") ||
    isSemVer(directory, true)
  );
}

function getVersion(version) {
  if (version === "master" || version === "next" || version === "latest") {
    return {
      version,
      branch: version,
    };
  }

  if (version.startsWith("_")) {
    return {
      version,
      branch: "topic",
      label: `#${version.substr(1)}`,
    };
  }

  return {
    version,
  };
}

const semVer = /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)$/;
const semVerIncludingPreRelease =
  /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-([0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))?$/;

function isSemVer(input, includePreRelease = false) {
  return input.match(includePreRelease ? semVerIncludingPreRelease : semVer);
}

function semverCompare(a, b) {
  a = a.split("-");
  b = b.split("-");
  var pa = a[0].split(".");
  var pb = b[0].split(".");
  for (var i = 0; i < 3; i++) {
    var na = Number(pa[i]);
    var nb = Number(pb[i]);
    if (na > nb) return 1;
    if (nb > na) return -1;
    if (!isNaN(na) && isNaN(nb)) return 1;
    if (isNaN(na) && !isNaN(nb)) return -1;
  }
  if (a[1] && b[1]) {
    return a[1] > b[1] ? 1 : a[1] < b[1] ? -1 : 0;
  }
  return !a[1] && b[1] ? 1 : a[1] && !b[1] ? -1 : 0;
}
```
